### ISSUE TITLE
Support for "forked" semver format <semver><symbol><semver>

### ISSUE THREAD

**Issue by hpe-ykoehler** [mojombo/semver#924]
Often we have to built software using multiple components from different sources.  We also often have to patch / fork the original package to fit within that software we are building.  But those patch may take time to be accepted upstream or may not fit with the original design direction of the upstream project.

There should be a way to track the original version and the forked version of a given package.

For example if I applied a breaking change to a package 1.2.3 version, I then want to express that but still keep the 1.2.3 version as the root, an example would be 1.2.3-2.0.0 but that '-' is certainly not the proper symbols to separate those.,but that is the idea where both version are kept and the "upstream" package version remains the first one since all the rest is based on it.  

As one upgrade the upstream package to 2.3.4 for example, you need to reconsider your patch, and they may or not be any breaking change anymore, so you would need to review that starting at 1.0.0 and adjust from there depending on the patch impact on the 2.3.4 version.

**Comment by ljharb** [comment]
If it's a fork, the project name should change.

**Comment by hpe-ykoehler** [comment]
Not really, I am not talking about public fork, but internal, where you upstream patches don't get accepted, you still need them and you eventually get them reworked to be accepted, and such, but all the in-between time you still need to track the package separately and keep track of the upstream project.  The goal of those forks isn't to start a new project but to adapt one to something else, and sometimes projects aren't flexible enough to allow for that without forking.

For example, in Gentoo, there were often cases where some package needed patches (which is basically like a fork) to adapt to gentoo environment, and since the package itself didn't offer the change requires as config, the code had to be changed.   In their case, they had to use versioning out of semver.

I think it would be nice for semver to acknowledge that such a scenario is actually quite frequent, as component gets adapted to different environment they often need adjustments, which sometimes can't be yet pushed upstream, and may take a couple of releases before the upstream project decide to add the require flexibility and such.

**Comment by ljharb** [comment]
That’s still a distinct project. I’d suggest keeping that metadata out of band, instead of coupling the version numbers of the public package with those of your internal fork.

It’s quite frequent but that doesn’t mean it’d advisable nor something that semver should accommodate.

**Comment by hpe-ykoehler** [comment]
Well, it is more the opposite, where the fork keeps track of the upstream project on which it is based (would not make sense for the public package to change its version to include forks).  Using a different name is hiding the information away, we then have to use other means to keep track of it, which means going away from semver, making semver less relevant.

**Comment by ljharb** [comment]
It doesn't hide it if you just slap a scope/prefix in front of it - an internal fork of `foo` at `acme` would be `@acme/foo` or `acme-foo`, problem solved.

**Comment by hpe-ykoehler** [comment]
The problem is not solved, if project foo is at 2.3.4, you then create @acme/foo at what version?

either you keep it, so @acme/foo 2.3.4, or you restart at @acme/foo 1.0.0

Then you evolve the code to add features if you re-used 2.3.4 you may now be at 4.1.2, if you restart from 1.0.0 you may be at 2.1.2.

During that time the upstream package has evolved and is now at 3.1.2

Now it may cause confusion since upstream had 3.1.2 and @acme/foo also had 3.1.2 at some time but the feature of the upstream foo in 3.1.2 isn't part of the @acme/foo 3.1.2, not even part of the 4.1.2, so we lost that data.

Then if you do merge the upstream foo 3.1.2 into the @acme/foo you then may increase @acme/foo to 5.0.0 and again you can lose that information of which feature set of foo your @acme/foo is based upon, which is a great amount of information, as it basically says All of the foo 3.1.2 API is or not part of the @acme/foo...

The fork I am talking about is not a fork in the sense that we are changing the design, but more about we are adapting the component to a specific env, yet the component remains the same, and its intent is to follow the upstream project.

And today each time I saw that pattern people deviated from the semver, making all tools out there build upon semver to require customization.  So that is why I think it would be worthwhile for semver to support such model so that tooling can be adjusted to match it and stop every project that use multiple components to go their own way.

The above scenario with 2 semver would be like

foo 2.3.4 -> @acme/foo 2.3.4-2.1.2
foo 3.1.2 -> @acme/foo 3.1.2-1.0.0

It better shows which upstream version is being talked about and how many changes were made to it.  Using '-' would not make sense I think, but one could easily see that 3.1.2-1.0.0 is a greater version than 2.3.4-2.1.2 and semver tool could be designed to support this model so that others can use it in their distro.

Pretty sure most distro that re-uses community package has to solve this problem on their own and likely they break semver and use some other means to express their own change to the package they imported/adapted.

**Comment by ljharb** [comment]
Then you can submit a PR to make things configurable.

An option that for some reason folks don't like to consider in these cases is "then you simply can't do that yet".

**Comment by hpe-ykoehler** [comment]
I think I already answered why that doesn't work.

**Comment by jwdonahue** [comment]
@hpe-ykoehler, I think this has been suggested in earlier proposals. I don't have time to find them right now.

I agree with @ljharb  that what you are trying to version is a separate product.  Here's the problem with your proposed solution:

- You fork at P 1.0.0 and subsequently release P 1.0.0@2.0.0 into your production system.
- I fork at P 1.0.0 and subsequently release P 1.0.0@2.0.0 into my production system.

Now we have a violation of the SemVer [clause 3](https://semver.org/#spec-item-3) which basically says "there can be only one".  You could argue that our respective production environments are isolated from each other, but my experience with this in large scale systems, is they are not.  I've seen multiple forks of various bits of code number in the dozens in some large scale development environments, where branch A of P has to be merged with B of P and the two branches don't have the same version sets of P.  Yes, each branch with multiple versions of P, but different sets! Yuk ;(.  Fixing build breaks in an environment with hundreds of branches and thousands of machines under these circumstances can be very painful. 

Packages have an unfortunate tendency to find their way into the wild. If your P v and my P v happen to be the same but with different bits in them, much confusion ensues. Better for us to rename our packages.  Whether our forks are ahead, behind or exactly at parity with the base product, our work is done in parallel to that of the base product owners, so there's always a risk of duplicate versions being applied to potentially different bits with the same name.

One solution is for you to always depend on your fork of P, ForkOfP.  For record keeping purposes, you can embed the version you forked from in the metadata, so your ForkOfP might have a version history along these lines:

1.0.0+P.2.1.0 // Initial fork without changes.
1.0.1+P.2.1.0 // You added your patch.
1.0.1+P.2.1.1 // Your patch was merged with base.
1.0.2+P.2.1.3 // You merged from base.
1.1.0+P.2.1.3 // You added a feature.
etc.

The downside to this is; if your product is consumed by others that also have a dependency on the base product:

- There's potential for some duplication of bits.
- There's some potential for incompatibility issues if the two versions rely on the same resources.
- There's some potential for namespace or symbol conflicts.

But your proposal already has those issues as well, and depending on the languages/environments involved there may or may not be straightforward solutions to them.  Diamond dependency issues have always been a hard problem.

The simplest thing to do is as @ljharb suggests and offer a PR to the base product owners.  If they are unable to respond reasonably quickly to your request, offer to pay them or volunteer your time to vet and accept PR's for them.

**Comment by hpe-ykoehler** [comment]
@jwdonahue I see your point about the fact that two distinct vendors could release the same package with the same name/version.  It actually can happen today as well, if distro one fix package P and create a 1.0.0-p1 and distro2 also apply a fix to package P and create a 1.0.0-p1 (if they use the same custom semver model).

It is clear that the distro will not "rename" a package because they add a patch to it, it is the same thing for vendors using open source, we do not rename a package because we add a patch to it.

So the solution of renaming a package makes no sense in the context because the idea is still that we want users of the vendor/distro to identify the function they need based on the package name they know..

Yet, to integrate there is pretty much often the need to apply patches, for all the reasons I already indicated (won't copy them back here).

I somehow feel that using a more explicit versioning system such as <upstream semver>-<integration-semver> would be more useful than <upstream-server>-<custom integration semver> and would convey better if a patch actually breaks something or not, without fixing the notion that there are now 2 distinct content with the same <packagename>-<version> because those would not be available together in the same environment, like today.

**Comment by ljharb** [comment]
@hpe-ykoehler note that things like a distro modifying the package without renaming it is also not really compatible with semver, in spirit if not in letter - only the maintainers decide what versions something has.

**Comment by ljharb** [comment]
That something is common doesn’t mean it’s at all advisable - designing a versioning system to accommodate it would be encouraging something bad.

**Comment by ljharb** [comment]
Distros can and should delay their release, or omit a package, if it’s not compatible.

**Comment by hpe-ykoehler** [comment]
@ljharb I guess we live in a different reality then.

**Comment by jwdonahue** [comment]
> Since there is no versioning spec in those scenarios we see people coming up with their own versioning scheme on top of semver. To me, that is an indication that there is a "need" that semver is not answering.

There are so many needs that SemVer does not, cannot, will not and the maintainers probably do not want to attempt to fill, that we could not delineate them all here. Not every tool chain has to follow SemVer, in fact, many packaging tools are not limited to SemVer only, precisely because it is not the end all and be all of versioning. 

Back in the 90's when I spent most of my time working embedded systems, it was not uncommon for us to install a base package (usually untar/zip something), then apply targeted patches to that base package for the particular board.  Pretty sure that's still a common thing.  I designed a packaging tool for the Windows build system about 10 years ago that made that sort of thing very easy to accomplish.  We used a manifest that defined what packages or parts thereof were laid down where on the drive(s), where they came from and the order in which they were laid down.  A "distro" in this sense was just a manifest that defined where base packages and patch packages come from and how they are laid out in storage.  No defiling of anybody's products or pretending to provide X when you're really providing Y, zero risk of content collisions.  And it requires manifests are versioned using SemVer, but content packages can use any unique name/version scheme the developers find useful (like build numbers or dates), though we were encouraging adoption of SemVer at the time.

That technique probably doesn't work for all things either, but it is another tool in the kit.

**Comment by jwdonahue** [comment]
> It actually can happen today as well, if distro one fix package P and create a 1.0.0-p1 and distro2 also apply a fix to package P and create a 1.0.0-p1 (if they use the same custom semver model).

That is not a SemVer compliant solution.  It's a violation of clause 3 and it doesn't make any sense due to the precedence rules.

**Comment by jwdonahue** [comment]
Add to the above that the package name is copyrighted, sometimes trademarked, and often not part of the OSS content. It is usually the intent of OSS publishers to grant rights in the source code and certain build outputs, not the name of the product. Some licenses are quite explicit in that regard.

**Comment by JuryA** [comment]
I believe this discussion is diverging from the core issue. Respectfully, I fail to see why a package should be renamed simply because it is a fork. In my experience, keeping a fork synchronized with the original project does not introduce any overwhelming complexity—especially with tools like Nix available. Renaming, therefore, seems more like an unnecessary complication rather than a useful or productive solution in this context.

The real problem, in my view, lies in how version metadata is handled—particularly when dealing with the infamous ‘+’ character. I regularly include additional information in the metadata segment of the version string. However, this quickly becomes a real blocker when the version already contains metadata. For instance, if the version is 1.2.3-bla+foo, figuring out whether a ‘+’ is already present and determining how to append further details (using a dot, hyphen, or some other workaround) becomes a surprisingly tedious and error-prone task. Frankly, it’s locking everyone out from proceeding cleanly, leading to some rather… creative and incompatible workarounds in version strings.

Given the reality of these constraints, I have to wonder why the rules around metadata are so tight. It doesn’t make much sense to me. Wouldn’t allowing multiple ‘+’ characters, or at least offering a clearer, more flexible standard for appending metadata, be a far more logical approach? It would certainly save us from having to invent these cumbersome workarounds just to deal with basic versioning practices.

**Comment by w6et** [comment]
One solution:{semver}.v{semver}.{classifier},such as ：
'3.1.2.v1.0.0','2.1.4.v1.0.0',
'3.1.2.v1.0.0.oss','3.1.2.v1.0.0.enterprise','3.1.2.v1.0.0.aws','3.1.2.v1.0.0.oca'
